### PR TITLE
Refactor XLA dialect registration

### PR DESCRIPTION
**Pull request by marbre**
* Implements an explicit registration of the XLA dialect
* Drops alwayslink flag from the XLA target within the CMake build configuration
* Drops the `iree_alwayslink` CMake module

### CODE REVIEW

**Comment by marbre** [comment]
I would like to see #1736 getting merged prior to this. I will rebase and complement this PR afterwards.

File: CMakeLists.txt [hunk 1]
```diff
@@ -75,7 +75,6 @@ include(iree_py_extension)
 include(iree_py_library)
 include(iree_py_test)
 include(iree_lit_test)
-include(iree_alwayslink)
```
Comment (GMNGeoffrey) [inline CMakeLists.txt:78]: :tada: !!!!!

File: iree/tools/init_xla_dialects.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace mlir {
+
+// This function should be called before creating any MLIRContext if one expect
+// all the possible dialects to be made available to the context automatically.
+inline void registerXLADialects() {
+  static bool init_once = []() {
+    registerDialect<mlir::xla_hlo::XlaHloDialect>();
+    registerDialect<mlir::xla_chlo::XlaHloClientDialect>();
+    registerDialect<mlir::xla_lhlo::XlaLhloDialect>();
```
Comment (GMNGeoffrey) [inline iree/tools/init_xla_dialects.h:36]: Do we use LHLO? I thought we didn't. If you want to leave it in for now and then try deleting it later, that's also fine with me :-)

File: iree/tools/BUILD [hunk 3]
```diff
@@ -139,13 +139,26 @@ cc_library(
...
+)
+
 cc_library(
     name = "iree_opt_main",
     srcs = ["opt_main.cc"],
     deps = [
         ":init_compiler_modules",
         ":init_passes_and_dialects",
         ":init_targets",
+        ":init_xla_dialects",
```
Comment (GMNGeoffrey) [inline iree/tools/BUILD:161]: We should be able to drop `xla_dialect_registration` below. If we can't, then I think this didn't work properly :-P

File: iree/tools/init_xla_dialects.h (hunk 2 above)
Comment (marbre) [reply iree/tools/init_xla_dialects.h:36]: This is just a copy&paste from TF's [dialect_registration.cc](https://github.com/tensorflow/tensorflow/blob/dfc190e78833f4d619657b908f9aeec8c4f09b0a/tensorflow/compiler/mlir/xla/ir/dialect_registration.cc#L24). Honestly, didn't checked if it is really used, but the `lhlo_ops` are at least included in the CMake tensorflow target. So, if we drop this we should do some more refactoring, maybe in a followup.

File: iree/tools/BUILD (hunk 3 above)
Comment (marbre) [reply iree/tools/BUILD:161]: Giving the CI a try :-)

File: CMakeLists.txt (hunk 1 above)
Comment (marbre) [reply CMakeLists.txt:78]: Be aware that this functions/module was only required for *external* deps. Internal targets are still linked as whole archive.

File: iree/tools/init_xla_dialects.h (hunk 2 above)
Comment (GMNGeoffrey) [reply iree/tools/init_xla_dialects.h:36]: :+1:

File: CMakeLists.txt (hunk 1 above)
Comment (GMNGeoffrey) [reply CMakeLists.txt:78]: Ah. Still nice though :-D

File: iree/tools/BUILD (hunk 3 above)
Comment (GMNGeoffrey) [reply iree/tools/BUILD:161]: Looks like it worked :-)

**Comment by GMNGeoffrey** [comment]
Were you going to drop the IREEVM dep on dialect registration? We can do that as a followup instead. Just wondering whether to start the merge

**Comment by marbre** [comment]
> Were you going to drop the IREEVM dep on dialect registration? We can do that as a followup instead. Just wondering whether to start the merge

Yeah, will do this in a follow up :) Anyway, it might be better to only merge #1736 or this one, because one of the two will need rebasing and some modifications to not break custom-opt and custom-translate.

**Comment by GMNGeoffrey** [comment]
Oh still waiting on someone else to approve that to be merged

**Comment by marbre** [comment]
This is now ready to pull :)

* XLA test passes registration is skipped and the dep/target will be removed/refactored in a followup as discussed on discord. 
* IREEVM's dep on XLA will be removed in a followup as well.

**Pull request closed without merging**